File: python/uwapi/events.py
from dataclasses import field
from typing import Callable, List
from .interop import *


class Events:
    _instance = None
    _connection_state_listeners: List[Callable[[UwConnectionStateEnum], None]] = []
    _game_state_listeners: List[Callable[[UwGameStateEnum], None]] = []
    _update_listeners: List[Callable[[bool], None]] = []
    _shootings_listeners: List[Callable[[UwShootingsArray], None]] = []
    _force_eliminated_listeners: List[Callable[[int], None]] = []
    _chat_listeners: List[Callable[[str, int, UwChatTargetFlags], None]] = []
    _map_state_listeners: List[Callable[[UwMapStateEnum], None]] = []

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def initialize(self) -> None:
        uw_interop.uwSetExceptionCallback(self._exception_callback)
        # uw_interop.uwSetLogCallback(self._log_callback)
        uw_interop.uwSetConnectionStateCallback(self._connection_state_callback)
        uw_interop.uwSetGameStateCallback(self._game_state_callback)
        uw_interop.uwSetUpdateCallback(self._update_callback)
        uw_interop.uwSetShootingsCallback(self._shootings_callback)
        uw_interop.uwSetForceEliminatedCallback(self._force_eliminated_callback)
        uw_interop.uwSetChatCallback(self._chat_callback)
        uw_interop.uwSetTaskCompletedCallback(self._task_completed_callback)
        uw_interop.uwSetMapStateCallback(self._map_state_callback)

    # ---------------------

    def on_connection_state(
        self, listener: Callable[[UwConnectionStateEnum], None]
    ) -> None:
        self._connection_state_listeners.append(listener)

    def on_game_state(self, listener: Callable[[UwGameStateEnum], None]) -> None:
        self._game_state_listeners.append(listener)

    def on_update(self, listener: Callable[[bool], None]) -> None:
        self._update_listeners.append(listener)

    def on_shootings(self, listener: Callable[[UwShootingsArray], None]) -> None:
        self._shootings_listeners.append(listener)

    def on_force_eliminated(self, listener: Callable[[int], None]) -> None:
        self._force_eliminated_listeners.append(listener)

    def on_chat(self, listener: Callable[[str, int, UwChatTargetFlags], None]) -> None:
        self._chat_listeners.append(listener)

    def on_map_state(self, listener: Callable[[UwMapStateEnum], None]) -> None:
        self._map_state_listeners.append(listener)

    # ---------------------

    def _exception_callback(self, message: str) -> None:
        print(f"exception: {message}")
        breakpoint()

    def _connection_state_callback(self, state: UwConnectionStateEnum) -> None:
        for listener in self._connection_state_listeners:
            listener(state)

    def _game_state_callback(self, state: UwGameStateEnum) -> None:
        for listener in self._game_state_listeners:
            listener(state)

    def _update_callback(self, stepping: bool) -> None:
        for listener in self._update_listeners:
            listener(stepping)

    def _shootings_callback(self, data: UwShootingsArray) -> None:
        for listener in self._shootings_listeners:
            listener(data)

    def _force_eliminated_callback(self, force: int) -> None:
        for listener in self._force_eliminated_listeners:
            listener(force)

    def _chat_callback(
        self, message: str, sender: int, flags: UwChatTargetFlags
    ) -> None:
        for listener in self._chat_listeners:
            listener(message, sender, flags)

    def _task_completed_callback(
        self, task_user_data: int, type: UwTaskTypeEnum
    ) -> None:
        pass  # todo

    def _map_state_callback(self, state: UwMapStateEnum) -> None:
        for listener in self._map_state_listeners:
            listener(state)


uw_events = Events()

File: csharp/uwapi/events.cs
using System;
using System.Runtime.InteropServices;

namespace Unnatural
{
    using ConnectionStateEnum = Interop.UwConnectionStateEnum;
    using GameStateEnum = Interop.UwGameStateEnum;
    using MapStateEnum = Interop.UwMapStateEnum;
    using ShootingsArray = Interop.UwShootingsArray;
    using ChatTargetFLags = Interop.UwChatTargetFlags;

    public class ChatMessage
    {
        public string Message;
        public uint Sender;
        public ChatTargetFLags Flags;
    }

    public static class Events
    {
        public static event EventHandler<ConnectionStateEnum> ConnectionStateChanged;
        public static event EventHandler<GameStateEnum> GameStateChanged;
        public static event EventHandler<MapStateEnum> MapStateChanged;
        public static event EventHandler<bool> Updating;
        public static event EventHandler<uint> ForceEliminated;
        public static event EventHandler<ChatMessage> ChatReceived;

        static readonly Interop.UwExceptionCallbackType ExceptionDelegate = new Interop.UwExceptionCallbackType(ExceptionCallback);
        static readonly Interop.UwConnectionStateCallbackType ConnectionStateDelegate = new Interop.UwConnectionStateCallbackType(ConnectionStateCallback);
        static readonly Interop.UwGameStateCallbackType GameStateDelegate = new Interop.UwGameStateCallbackType(GameStateCallback);
        static readonly Interop.UwMapStateCallbackType MapStateDelegate = new Interop.UwMapStateCallbackType(MapStateCallback);
        static readonly Interop.UwUpdateCallbackType UpdateDelegate = new Interop.UwUpdateCallbackType(UpdateCallback);
        static readonly Interop.UwShootingsCallbackType ShootingsDelegate = new Interop.UwShootingsCallbackType(ShootingsCallback);
        static readonly Interop.UwForceEliminatedCallbackType ForceEliminatedDelegate = new Interop.UwForceEliminatedCallbackType(ForceEliminatedCallback);
        static readonly Interop.UwChatCallbackType ChatDelegate = new Interop.UwChatCallbackType(ChatCallback);

        static void ExceptionCallback([MarshalAs(UnmanagedType.LPStr)] string message)
        {
            Console.WriteLine("exception: " + message);
            if (System.Diagnostics.Debugger.IsAttached)
                System.Diagnostics.Debugger.Break();
        }

        static void ConnectionStateCallback(ConnectionStateEnum state)
        {
            if (ConnectionStateChanged != null)
                ConnectionStateChanged(null, state);
        }

        static void GameStateCallback(GameStateEnum state)
        {
            if (GameStateChanged != null)
                GameStateChanged(null, state);
        }

        static void MapStateCallback(MapStateEnum state)
        {
            if (MapStateChanged != null)
                MapStateChanged(null, state);
        }

        static void UpdateCallback(bool stepping)
        {
            if (Updating != null)
                Updating(null, stepping);
        }

        static void ShootingsCallback(ref ShootingsArray data)
        {
            // todo decode shooting
        }

        static void ForceEliminatedCallback(uint force)
        {
            if (ForceEliminated == null)
                return;
            ForceEliminated(null, force);
        }

        static void ChatCallback(string msg, uint sender, ChatTargetFLags flags)
        {
            if (ChatReceived == null)
                return;
            ChatMessage c = new ChatMessage();
            c.Message = msg;
            c.Sender = sender;
            c.Flags = flags;
            ChatReceived(null, c);
        }

        static Events()
        {
            AppDomain.CurrentDomain.ProcessExit += Destructor;
            Interop.uwInitialize(Interop.UW_VERSION);
            Interop.uwInitializeConsoleLogger();

            Interop.uwSetExceptionCallback(ExceptionDelegate);
            Interop.uwSetConnectionStateCallback(ConnectionStateDelegate);
            Interop.uwSetGameStateCallback(GameStateDelegate);
            Interop.uwSetMapStateCallback(MapStateDelegate);
            Interop.uwSetUpdateCallback(UpdateDelegate);
            Interop.uwSetShootingsCallback(ShootingsDelegate);
            Interop.uwSetForceEliminatedCallback(ForceEliminatedDelegate);
            Interop.uwSetChatCallback(ChatDelegate);

            // make sure that others register their callbacks too (call any method)
            Prototypes.All();
            Map.Positions();
            World.Entities();
            UwapiTasks.Init();
        }

        static void Destructor(object sender, EventArgs e)
        {
            Interop.uwDeinitialize();
        }
    }

    public static class UwapiTasks
    {
        public static ulong InsertTask(Action a)
        {
            ulong i = index++;
            actions.Add(i, a);
            return i;
        }

        static ulong index = 1;
        static System.Collections.Generic.Dictionary<ulong, Action> actions = new System.Collections.Generic.Dictionary<ulong, Action>();
        static readonly Interop.UwTaskCompletedCallbackType TaskCompletedDelegate = new Interop.UwTaskCompletedCallbackType(TaskCompleted);

        static void TaskCompleted(ulong taskUserData, Interop.UwTaskTypeEnum type)
        {
            if (type != Interop.UwTaskTypeEnum.None)
            {
                Action a;
                if (actions.TryGetValue(taskUserData, out a))
                    a();
            }
            actions.Remove(taskUserData);
        }

        static UwapiTasks()
        {
            Interop.uwSetTaskCompletedCallback(TaskCompletedDelegate);
        }

        public static void Init()
        { }
    }
}
